crates/trie/trie/src/proof_v2/mod.rs

```diff
@@ -570,27 +646,450 @@ where
+            let (nibble, short_key) = self.push_new_branch(cached_path);
+            (nibble, short_key.len())
+        } else {
+            // If there is a parent branch but its `state_mask` bit for this branch is not set
+            // then we can simply calculate the `ext_len` based on the difference of each, minus
+            // 1 to account for the nibble in the `state_mask`.
+            (cached_branch_nibble, cached_path.len() - self.branch_path.len() - 1)
+        };
+
+        // `commit_last_child` relies on the last set bit of the parent branch's `state_mask` to
+        // determine the path of the last child on the `child_stack`. Since we are about to
+        // change that mask we need to commit that last child first.
+        self.commit_last_child(targets)?;
+
+        // When pushing a new branch we need to set its child nibble in the `state_mask` of
+        // its parent, if there is one.
+        if let Some(parent_branch) = self.branch_stack.last_mut() {
+            parent_branch.state_mask.set_bit(cached_branch_nibble);
+        }
+
+        // Finally update the `branch_path` and push the new branch.
+        self.branch_path = cached_path;
+        self.branch_stack.push(Self::new_from_cached_branch(cached_branch, ext_len as u8));
+
+        trace!(
+            target: TRACE_TARGET,
+            branch=?self.branch_stack.last(),
+            branch_path=?self.branch_path,
+            "Pushed cached branch",
+        );
+
+        Ok(())
+    }
+
+    /// Accepts the current state of both hashed and trie cursors, and determines the next range of
+    /// hashed keys which need to be processed using [`Self::push_leaf`].
+    ///
+    /// This method will use cached branch node data from the trie cursor to skip over all possible
+    /// ranges of keys, to reduce computation as much as possible.
+    ///
+    /// # Returns
+    ///
+    /// - `None`: No more data to process, finish computation
+    ///
+    /// - `Some(lower, None)`: Indicates to call `push_leaf` on all keys starting at `lower`, with
+    ///   no upper bound. This method won't be called again after this.
+    ///
+    /// - `Some(lower, Some(upper))`: Indicates to call `push_leaf` on all keys starting at `lower`,
+    ///   up to but excluding `upper`, and then call this method once done.
+    #[instrument(target = TRACE_TARGET, level = "trace", skip_all)]
+    fn next_uncached_key_range(
+        &mut self,
+        targets: &mut TargetsIter<impl Iterator<Item = Nibbles>>,
+        trie_cursor_state: &mut TrieCursorState,
+        hashed_key_current_path: Option<Nibbles>,
+    ) -> Result<Option<(Nibbles, Option<Nibbles>)>, StateProofError> {
+        // All trie data prior to the current cached branch, if any, has been computed. Any branches
+        // which were under-construction previously, and which are not on the same path as this
+        // cached branch, can be assumed to be completed; they will not have any further keys added
```

Comment:
```suggestion
        // Pop any under-construction branches that are now complete.
        // All trie data prior to the current cached branch, if any, has been computed. Any branches
        // which were under-construction previously, and which are not on the same path as this
        // cached branch, can be assumed to be completed; they will not have any further keys added.
```

Found that it was slightly hard to follow this. Would help i think if we mention what the purpose is and followed by the why.